libsignal-service/src/provisioning/manager.rs

```diff
@@ -401,3 +433,126 @@ impl<P: PushService> LinkingManager<P> {
         Ok(())
     }
 }
+
+#[derive(Debug)]
+pub struct DeviceName {
+    ephemeral_public: PublicKey,
+    synthetic_iv: Vec<u8>,
+    ciphertext: Vec<u8>,
+}
+
+impl TryInto<crate::proto::DeviceName> for DeviceName {
+    type Error = libsignal_protocol::SignalProtocolError;
+
+    fn try_into(self) -> Result<crate::proto::DeviceName, Self::Error> {
+        Ok(crate::proto::DeviceName {
+            ephemeral_public: Some(
+                self.ephemeral_public.public_key_bytes()?.to_vec(),
+            ),
+            synthetic_iv: Some(self.synthetic_iv.to_vec()),
+            ciphertext: Some(self.ciphertext),
+        })
+    }
+}
+
+fn calculate_hmac256(
+    mac_key: &[u8],
+    ciphertext: &[u8],
+) -> Result<Vec<u8>, ServiceError> {
+    let mut mac = Hmac::<Sha256>::new_from_slice(mac_key)
+        .map_err(|_| ServiceError::MacError)?;
+    mac.update(&ciphertext);
+    Ok(mac.finalize().into_bytes().to_vec())
+}
+
+pub fn encrypt_device_name<R: rand::Rng + rand::CryptoRng>(
+    csprng: &mut R,
+    device_name: &str,
+    identity_public: &PublicKey,
+) -> Result<DeviceName, ProvisioningError> {
+    let plaintext = device_name.as_bytes().to_vec();
+    let ephemeral_key_pair = KeyPair::generate(csprng);
+
+    let master_secret = ephemeral_key_pair
+        .private_key
+        .calculate_agreement(identity_public)?;
+
+    let key1 = calculate_hmac256(&master_secret, b"auth")?;
+    let mut synthetic_iv = calculate_hmac256(&key1, &plaintext)?;
```

Comment:
Doesn't seem necessary to allocate `plaintext` as `Vec<u8>`.

```suggestion
    let mut synthetic_iv = calculate_hmac256(&key1, device_name.as_bytes())?;
```

and remove the plaintext allocating lines.